presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Rule.java

```diff
@@ -46,4 +57,9 @@ default boolean isEnabled(Session session)
 
         Session getSession();
     }
+
+    interface Result
+    {
+        Result transformTo(PlanNode planNode);
```

Comment:
Exactly! This what `IterativeOptimizer` currently allows. I agree that thee two kind of methods you suggested  (and even some more) will be eventually needed, but for now I would prefer not to pretend that something is possible while it is not.

Regarding cost related rules I would prefer to port them one by one to observe how optimizer is going to behave. So I prefer to leave them as they are for now. Notice that is a plain refactor commit which means that there is no change in semantics of anything. 

If you want I can rename this to `replace`.